### ISSUE TITLE
Serious concerns about StreamMind paper and its implementation

### ISSUE THREAD

**Issue by canornot** [xinding-sys/StreamMind#9]
We carefully studied the StreamMind paper and its open-source code. Using publicly available data — SoccerNet and VideoLLaMA2 weights as pretrained models — we attempted to reproduce and retrain StreamMind. Unfortunately, despite our best efforts, the experimental results were disappointing. Upon close inspection of the codebase, we identified several critical issues in the implementation:

---

## ❶ Ground Truth Is Directly Fed into the Model

In the official evaluation script ([Evaluate.sh](https://github.com/xinding-sys/StreamMind/blob/main/scripts/eval/evaluation.sh)), the authors feed the ground truth captions directly into the model. 

Specifically, the evaluation uses the model’s `forward` function instead of `generate`:  👉 [Code reference](https://github.com/xinding-sys/StreamMind/blob/main/streammind/eval/inference_video_ego4d_stream_parallel_new.py#L184)

The input data is processed using the original training pipeline:  👉 [Data pipeline](https://github.com/xinding-sys/StreamMind/blob/main/streammind/eval/inference_video_ego4d_stream_parallel_new.py#L163)

The key issue is that in the `forward` function, **the ground truth caption is fed into the model along with the query**:  👉 [Code reference](https://github.com/xinding-sys/StreamMind/blob/main/data/soccer_data.py#L69)

This means the model effectively “sees the answer” during evaluation.

Using the authors’ inference code, we observed that the generated captions are nearly identical to the ground truth:

<img width="1177" height="261" alt="Image" src="https://github.com/user-attachments/assets/ae16b022-4af6-4b2c-a7d4-1c0c8b00a1d5" />

The final evaluation scores also closely match those reported in the paper (minor differences due to LoRA training and GPU constraints):

<img width="1669" height="405" alt="Image" src="https://github.com/user-attachments/assets/71b93e0b-e3b8-45ea-96ac-85b786007632" />

However, when we removed the ground truth or replaced it with a simple, unrelated caption like “I love you,” the model’s output still aligned closely with the modified caption — even though the video input remained unchanged:

<img width="695" height="201" alt="Image" src="https://github.com/user-attachments/assets/8c81f020-46ed-4379-b309-eda171b4e113" />

This behavior — where changing the ground truth alters the model output — is deeply concerning.

---

## ❷ Abnormal Behavior in Generate Mode

We further tested the model using the `generate` function and found that its output barely changes with different video inputs:

- In [`inference_video_score.py`](https://github.com/xinding-sys/StreamMind/blob/main/streammind/eval/inference_video_score.py), we sampled 40 frames every 200 seconds and fed them into the model. The generated captions were completely identical:

<img width="1355" height="239" alt="Image" src="https://github.com/user-attachments/assets/eeb44228-652d-4b44-9b79-7b25697bd6bb" />

- In [`inference_video_score_stream.py`](https://github.com/xinding-sys/StreamMind/blob/main/streammind/eval/inference_video_score_stream.py), we streamed video frames every 0.5 seconds and asked the model to decide whether to “open” or “close” a gate. The model consistently predicted “open” every second, with identical descriptions — far from the dynamic behavior shown in the demo video:

<img width="1373" height="295" alt="Image" src="https://github.com/user-attachments/assets/8d965681-4fd3-4ea2-a58f-ecbd22a8cdd6" />

We suspect the root cause lies in how StreamMind handles token compression. The model averages multiple input tokens into a single token:  
👉 [Token compression code](https://github.com/xinding-sys/StreamMind/blob/main/streammind/model/multimodal_projector/builder.py#L405)

Unlike VideoLLaMA2’s STC_Connector, which maps multi-frame video into rich token sequences, StreamMind’s Mamba-based architecture compresses the input too aggressively, resulting in significant loss of video information and poor content extraction.

---

## 📉 Conclusion

While the paper’s Table 1 and Table 2 report impressive metrics, these results are based on `forward`-mode evaluation with ground truth fed into the model — not actual `generate`-mode performance.

When tested properly using `generate`, the model’s captions and cognition gate outputs barely respond to changes in video input. This raises serious questions — **could this be why the authors have not released the model weights?**

We welcome an official response from the authors, and invite others who have attempted to reproduce the paper — or plan to attend ICCV 2025 — to join the discussion.

**Comment by xinding-sys** [comment]
First of all, thank you for your interest. Regarding the issues you raised, we believe there may be some misunderstandings as well as differences in the reproduction process. Both generation and forward are crucial for the model. This is why we adopt BLEU-1, BLEU-4, METEOR, and ROUGE to evaluate generation, while following established evaluation systems such as videollm-online with perplexity (PPL) and correction metrics for forward. This is consistent with prior work. Specifically, the evaluation of forward and generation reflects the model’s ability in the two complementary dimensions of “understanding–prediction” and “expression–generation.” Only by combining both can we comprehensively assess the model’s performance on complex tasks.

Second, regarding your comment that “StreamMind directly averages multiple input tokens into a single token,” this may be a misunderstanding. When calling the LLM, past tokens are still required, not just a single compressed token. That said, StreamMind indeed has limitations—the compression of single-frame information remains somewhat aggressive. While this is not a bottleneck on event-level benchmarks such as SoccerNet, balancing efficiency and compression ratio is still an open direction for future discussion and improvement.

As for open-sourcing the model weights, this is tied to our ongoing commercialization plan. I cannot decide this individually, but the release is expected to happen later this year.

Finally, I sincerely apologize for the delayed response—I just noticed your message. Due to the rushed release, some redundant code has not yet been fully organized, which may have caused confusion and inconvenience. I will further clean and refine the repository after completing my internship in the near future. Regarding the anomalies in your results, I believe they may be related to certain overlooked or misaligned details during training, as there are indeed many aspects that require careful handling. If you would like timely support in your reproduction efforts, please feel free to reach me via WeChat (vx) or email at xinding64@mail.ustc.edu.cn. I would be glad to provide any assistance I can, and I warmly welcome further exchanges!

<img width="313" height="308" alt="Image" src="https://github.com/user-attachments/assets/36e29a00-2b4d-416e-9b0c-858eb2dc0124" />

**Comment by canornot** [comment]
Grateful to the author for patiently taking the time to communicate with us directly. We are most likely still limited by the size of our GPU, which forced us to use LoRA during training and restrict the batch size to only 1. As a result, the model was unable to reach the global minimum, leading to suboptimal results. For those attempting to reproduce our work, sufficient GPU memory is still necessary. Many thanks again to the author for the patient and detailed explanation!
